pkg/kgo/pools.go

```diff
@@ -0,0 +1,207 @@
+	// just have extra data slices put back into your pool that you never
+	// created.
+	GetDecompressBytes(compressed []byte, codec CompressionCodecType) []byte
+	// PutDecompressBytes puts a slice of that was used for decompression
+	// back into the pool. The slice is zeroed before it is put back.
+	PutDecompressBytes([]byte)
+}
+
+// PoolKRecords is a pool that returns a slice that raw kmsg.Record's are
+// decoded into.
+type PoolKRecords interface {
+	// GetKRecords returns a slice with capacity n.
+	GetKRecords(n int) []kmsg.Record
+	// PutKRecords puts a slice back into the pool.
+	PutKRecords([]kmsg.Record)
+}
+
+// PoolRecords is a pool that returns a slice of Record's.
+type PoolRecords interface {
+	// GetRecords returns a slice with capacity n.
+	GetRecords(n int) []Record
+	// PutRecords puts a slice back into the pool.
+	PutRecords([]Record)
+}
+
+func strp(s string) *string { return &s }
+
+var ctxRecRecycle = strp("rec-recycle")
+
+const recSize = unsafe.Sizeof(Record{})
+
+// krecs and recs are *always* non-nil. If the user did not give us pooled
+// values, we create our own. The user could only pool one of the three, that's
+// fine.
+func recordPoolsCtx(pools []Pool, decompressBytes []byte, krecs []kmsg.Record, recs []Record) (*recordPools, context.Context) {
+	if len(recs) == 0 { // ...just in case
+		return nil, nil
+	}
+	p := &recordPools{
+		pools:           pools,
+		decompressBytes: decompressBytes,
+		krecs:           krecs,
+		recs:            recs,
+	}
+	return p, context.WithValue(context.Background(), ctxRecRecycle, p)
+}
+
+// Recycle "recycles" this record if it was taken from a pool, and frees its
+// attachment to any underlying pooled slices. If the pooled slice no longer
+// has any records attached, the slices are put back into their pools.
+//
+// This method is only relevant if you are using the [WithPools] option.
+//
+// NOTE: It is invalid to continue using the record after calling recycle;
+// doing so may result in corruption and data races. If you use
+// PoolDecompressBytes, you cannot continue to use a shallow copy of any
+// fields, you must clone them!
+func (r *Record) Recycle() {
+	if r.Context == nil {
+		return
+	}
+	v := r.Context.Value(ctxRecRecycle)
+	if v == nil {
+		return
+	}
+	*r = Record{} // prevent the Record from hanging onto anything; *kmsg.Record is already cleared during processing
+	ps := v.(*recordPools)
+
+	r0 := &ps.recs[0]
+
+	idx := int((uintptr(unsafe.Pointer(r)) - uintptr(unsafe.Pointer(r0))) / recSize) //nolint:gosec // unsafe rule (3) roughly (we don't need to convert back)
+	if idx < 0 || idx >= len(ps.recs) {
+		panic(fmt.Sprintf("recycling a record and the index %v is invalid (max pool len %v)", idx, len(ps.recs)))
+	}
+	if &ps.recs[idx] != r {
+		panic("sanity check pointer comparison index for recycling record is not the same!")
+	}
+
+	rem := ps.n.Add(-1)
+	if rem > 0 {
+		return
+	}
+
+	// Reset the capacity of slices to max to ensure we zero things and
```

Comment:
nitpick on the comment: this only resets the length of the slices, their capacity is still the same. Also the sentence in the comment isn't complete